builtin/credential/cert/path_login.go

```diff
@@ -170,49 +186,56 @@ func (b *backend) verifyCredentials(req *logical.Request) (*ParsedCert, *logical
 	}
 
-	validChain := b.checkForValidChain(trustedChains)
-	if !validChain {
-		return nil, logical.ErrorResponse(
-			"no chain containing non-revoked certificates could be found for this login certificate",
-		), nil
+	// Search for a ParsedCert that intersects with the validated chains and any additional constraints
+	matches := make([]*ParsedCert, 0)
+	for _, trust := range trusted { // For each ParsedCert in the config
+		for _, tCert := range trust.Certificates { // For each certificate in the entry
+			for _, chain := range trustedChains { // For each root chain that we matched
+				for _, cCert := range chain { // For each cert in the matched chain
+					if tCert.Equal(cCert) && // ParsedCert intersects with matched chain
+						b.matchesConstraints(connState.PeerCertificates[0], chain, trust) { // validate client cert + matched chain against the config
+						// Add the match to the list
+						matches = append(matches, trust)
+					}
+				}
+			}
+		}
+	}
+
+	// Fail on no matches
+	if len(matches) == 0 {
+		return nil, logical.ErrorResponse("no chain matching all constraints could be found for this login certificate"), nil
 	}
 
-	// Match the trusted chain with the policy
-	return b.matchPolicy(trustedChains, trusted), nil, nil
+	// Return the first matching entry (for backwards compatibility, we continue to just pick one if multiple match)
+	return matches[0], nil, nil
 }
 
-// matchNonCAPolicy is used to match the client cert with the registered non-CA
-// policies to establish client identity.
-func (b *backend) matchNonCAPolicy(clientCert *x509.Certificate, trustedNonCAs []*ParsedCert) *ParsedCert {
-	for _, trustedNonCA := range trustedNonCAs {
-		tCert := trustedNonCA.Certificates[0]
-		if tCert.SerialNumber.Cmp(clientCert.SerialNumber) == 0 && bytes.Equal(tCert.AuthorityKeyId, clientCert.AuthorityKeyId) {
-			return trustedNonCA
+func (b *backend) matchesConstraints(clientCert *x509.Certificate, trustedChain []*x509.Certificate, config *ParsedCert) bool {
+	nameMatched := (config.Entry.RequiredName == "")
+	if config.Entry.RequiredName != "" {
+		if glob.Glob(config.Entry.RequiredName, clientCert.Subject.CommonName) {
+			nameMatched = true
 		}
-	}
-	return nil
-}
 
-// matchPolicy is used to match the associated policy with the certificate that
-// was used to establish the client identity.
-func (b *backend) matchPolicy(chains [][]*x509.Certificate, trusted []*ParsedCert) *ParsedCert {
-	// There is probably a better way to do this...
-	for _, chain := range chains {
-		for _, trust := range trusted {
-			for _, tCert := range trust.Certificates {
-				for _, cCert := range chain {
-					if tCert.Equal(cCert) {
-						return trust
-					}
-				}
+		for _, name := range clientCert.DNSNames {
+			if glob.Glob(config.Entry.RequiredName, name) {
+				nameMatched = true
+			}
+		}
+
+		for _, name := range clientCert.EmailAddresses {
+			if glob.Glob(config.Entry.RequiredName, name) {
+				nameMatched = true
 			}
```

Comment:
Technically IP SANs are IP SANames :-) Hence why I thought about it. I guess let's leave it off for now; it can always be added later, either as string checks or subnet checks.